src/ModelContextProtocol.AspNetCore/StreamableHttpHandler.cs

```diff
@@ -0,0 +1,145 @@
+﻿using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Http.Features;
+using Microsoft.AspNetCore.WebUtilities;
+using Microsoft.Extensions.Hosting;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using ModelContextProtocol.Protocol.Messages;
+using ModelContextProtocol.Protocol.Transport;
+using ModelContextProtocol.Server;
+using ModelContextProtocol.Utils.Json;
+using System.Collections.Concurrent;
+using System.Security.Cryptography;
+
+namespace ModelContextProtocol.AspNetCore;
+
+internal sealed class StreamableHttpHandler(
+    IOptions<McpServerOptions> mcpServerOptionsSnapshot,
+    IOptionsFactory<McpServerOptions> mcpServerOptionsFactory,
+    IOptions<HttpServerTransportOptions> httpMcpServerOptions,
+    IHostApplicationLifetime hostApplicationLifetime,
+    ILoggerFactory loggerFactory)
+{
+
+    private readonly ConcurrentDictionary<string, HttpMcpSession> _sessions = new(StringComparer.Ordinal);
+    private readonly ILogger _logger = loggerFactory.CreateLogger<StreamableHttpHandler>();
+
+    public async Task HandleRequestAsync(HttpContext context)
+    {
+        if (context.Request.Method == HttpMethods.Get)
+        {
+            await HandleSseRequestAsync(context);
+        }
+        else if (context.Request.Method == HttpMethods.Post)
+        {
+            await HandleMessageRequestAsync(context);
+        }
+        else
+        {
+            context.Response.StatusCode = StatusCodes.Status405MethodNotAllowed;
+            await context.Response.WriteAsync("Method Not Allowed");
+        }
+    }
+
+    public async Task HandleSseRequestAsync(HttpContext context)
+    {
+        // If the server is shutting down, we need to cancel all SSE connections immediately without waiting for HostOptions.ShutdownTimeout
+        // which defaults to 30 seconds.
+        using var sseCts = CancellationTokenSource.CreateLinkedTokenSource(context.RequestAborted, hostApplicationLifetime.ApplicationStopping);
+        var cancellationToken = sseCts.Token;
+
+        var response = context.Response;
+        response.Headers.ContentType = "text/event-stream";
+        response.Headers.CacheControl = "no-cache,no-store";
+
+        // Make sure we disable all response buffering for SSE
+        context.Response.Headers.ContentEncoding = "identity";
+        context.Features.GetRequiredFeature<IHttpResponseBodyFeature>().DisableBuffering();
+
+        var sessionId = MakeNewSessionId();
+        await using var transport = new SseResponseStreamTransport(response.Body, $"message?sessionId={sessionId}");
```

Comment:
Were you saying something about wanting to data protect session IDs?

Reply:
I think I still might have to do this for stateless requests similar to what we do for antiforgery, since we won't be able to rely on the ConcurrentDictionary to track if things like the user ID changed if the client sends requests to multiple servers, but I haven't got there yet.